// eslint-plugin-vue-scoped-css/dist/styles/utils/css-nodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isVCSSAtRule = isVCSSAtRule;
exports.isVCSSStyleRule = isVCSSStyleRule;
exports.isVCSSStyleSheet = isVCSSStyleSheet;
exports.isVCSSDeclarationProperty = isVCSSDeclarationProperty;
exports.isVCSSComment = isVCSSComment;
exports.isVCSSContainerNode = isVCSSContainerNode;
exports.hasSelectorNodes = hasSelectorNodes;
const selectors_1 = require("./selectors");
function isVCSSAtRule(node) {
    return (node === null || node === void 0 ? void 0 : node.type) === "VCSSAtRule";
}
function isVCSSStyleRule(node) {
    return (node === null || node === void 0 ? void 0 : node.type) === "VCSSStyleRule";
}
function isVCSSStyleSheet(node) {
    return (node === null || node === void 0 ? void 0 : node.type) === "VCSSStyleSheet";
}
function isVCSSDeclarationProperty(node) {
    return (node === null || node === void 0 ? void 0 : node.type) === "VCSSDeclarationProperty";
}
function isVCSSComment(node) {
    return (node === null || node === void 0 ? void 0 : node.type) === "VCSSComment" || (node === null || node === void 0 ? void 0 : node.type) === "VCSSInlineComment";
}
function isVCSSContainerNode(node) {
    return (isVCSSAtRule(node) ||
        isVCSSStyleRule(node) ||
        isVCSSStyleSheet(node) ||
        (node === null || node === void 0 ? void 0 : node.type) === "VCSSUnknown");
}
function hasSelectorNodes(node) {
    if (isVCSSStyleRule(node) || (0, selectors_1.isNestingAtRule)(node)) {
        return true;
    }
    return false;
}
